src/coreclr/jit/optimizebools.cpp

```diff
@@ -0,0 +1,1416 @@
+            // t1:c1==0 t2:c2!=0 ==> Branch to BX if both values are non-0
+            // So we will branch to BX if (c1&c2)!=0
+
+            foldOp = GT_AND;
+            cmpOp  = GT_NE;
+        }
+        else if (m_testInfo1.compTree->gtOper == GT_LT && m_testInfo2.compTree->gtOper == GT_GE)
+        {
+            // t1:c1<0 t2:c2>=0 ==> Branch to BX if both values >= 0
+            // So we will branch to BX if (c1|c2)>=0
+
+            foldOp = GT_OR;
+            cmpOp  = GT_GE;
+        }
+        else if (m_testInfo1.compTree->gtOper == GT_GE)
+        {
+            return false;
+        }
+        else if (m_testInfo1.compTree->gtOper == GT_NE && m_testInfo2.compTree->gtOper == GT_EQ)
+        {
+            // t1:c1!=0 t2:c2==0 ==> Branch to BX if both values are 0
+            // So we will branch to BX if (c1|c2)==0
+
+            foldOp = GT_OR;
+            cmpOp  = GT_EQ;
+        }
+        else
+        {
+            return false;
+        }
+    }
+
+    // Anding requires both values to be 0 or 1
+
+    if ((foldOp == GT_AND) && (!m_testInfo1.isBool || !m_testInfo2.isBool))
+    {
+        return false;
+    }
+
+    //
+    // Now update the trees
+    //
+
+    m_foldOp   = foldOp;
+    m_foldType = foldType;
+    m_cmpOp    = cmpOp;
+
+    optOptimizeBoolsUpdateTrees();
+
+#ifdef DEBUG
+    if (m_comp->verbose)
+    {
+        printf("Folded %sboolean conditions of " FMT_BB " and " FMT_BB " to :\n", m_c2->OperIsLeaf() ? "" : "non-leaf ",
+               m_b1->bbNum, m_b2->bbNum);
+        m_comp->gtDispStmt(s1);
+        printf("\n");
+    }
+#endif
+
+    // Return true to continue the bool optimization for the rest of the BB chain
+    return true;
+}
+
+//-----------------------------------------------------------------------------
+//  FindCompareChain:  Check if the given condition is a compare chain.
+//
+// Arguments:
+//      condition:        Condition to check.
+//      isTestCondition:  Returns true if condition is a EQ/NE(AND(...),0) but is not a compare chain.
+//
+//  Returns:
+//      true if chain optimization is a compare chain.
+//
+//  Assumptions:
+//      m_b1 and m_b2 are set on entry.
+//
+
+inline bool OptBoolsDsc::FindCompareChain(GenTree* condition, bool* isTestCondition)
+{
+    GenTree* condOp1 = condition->gtGetOp1();
+    GenTree* condOp2 = condition->gtGetOp2();
+
+    *isTestCondition = false;
+
+    if (condition->OperIs(GT_EQ, GT_NE) && condOp2->IsIntegralConst() && condOp2->AsIntCon()->IconValue() == 0)
+    {
+        // Found a test condition. Does it contain a compare chain?
+
+        // Only test that the second operand of AND ends with a compare operation, as this will be
+        // the condition the new link in the chain will connect with.
+        // We are allowing for the first operand of the not be a valid chain, as this would require
+        // a full recursive search through the children.
+        if (condOp1->OperIs(GT_AND) && condOp1->gtGetOp2()->OperIsCmpCompare())
```

Comment:
```suggestion
        if (condOp1->OperIs(GT_AND) && condOp1->gtGetOp2()->OperIsCmpCompare() && varTypeIsIntegralOrI(condOp1->gtGetOp2()->gtGetOp1()))
```

Should this check be symmetrical? What if the relop is the other operand?

Reply:
Here we want to check for `AND(XXX,CMP)` where XXX is either a compare or another `AND` of conditions.

Crucially op2 needs to be a compare, as this is the current end of the chain. When the chain is extended, the new condition will connect op2 when the CCMPs are generated.

If op1 was something else entirely, eg `AND(SUB, CMP)`, then this code still allows it. The resulting code would probably be non optimal, but I'm not seeing any cases where that happens (I added some assert checks and did an spmi replay). Maybe a slightly better version would just be to ensure op1 is either a compare or `AND`.

Consider `AND(CMP, SUB)`. That's not going to work as the `SUB` will be generated after the compare. I guess we could swap op1 and op2 in that case. But, again there won't be any instances of this. 

Meanwhile `AND(CMP, AND)`, where the second `AND` is another chain. Adding some asserts, this does fire. So maybe I do need to recursive check the second AND.

Reply:
My main worry would be having one operand that is a floating-point comparison and one operand that is an integer comparison. I suppose that only becomes a problem if you add my suggested change, but I assume we do want that check since we cannot emit `ccmp` for fp comparisons.